be/op.c

```diff
@@ -294,21 +316,135 @@ M0_INTERNAL void m0_be_op_wait(struct m0_be_op *op)
+	M0_PRE(op->bo_kind == M0_BE_OP_KIND_NORMAL);
+	m0_be_op_active(op);
+	m0_be_op_lock(op);
+	op->bo_kind = op_kind;
+	m0_be_op_unlock(op);
+}
+
+M0_INTERNAL void m0_be_op_make_set_and(struct m0_be_op *op)
+{
+	M0_ENTRY("op=%p", op);
+	be_op_set_make(op, M0_BE_OP_KIND_SET_AND);
+	M0_LEAVE("op=%p", op);
+}
+
+M0_INTERNAL void m0_be_op_make_set_or(struct m0_be_op *op)
+{
+	M0_ENTRY("op=%p", op);
+	be_op_set_make(op, M0_BE_OP_KIND_SET_OR);
+	M0_LEAVE("op=%p", op);
+}
+
 M0_INTERNAL void m0_be_op_set_add(struct m0_be_op *parent,
 				  struct m0_be_op *child)
 {
-	/* lock order here and in be_op_state_change() should be the same */
+	M0_ENTRY("parent=%p child=%p", parent, child);
+	/*
+	 * Lock order here, in be_op_done() and in
+	 * m0_be_op_set_add_finish() should be the same.
+	 */
 	m0_be_op_lock(child);
 	m0_be_op_lock(parent);
 
+	M0_ASSERT(M0_IN(parent->bo_kind, (M0_BE_OP_KIND_SET_AND,
+	                                  M0_BE_OP_KIND_SET_OR)));
 	M0_ASSERT(parent->bo_sm.sm_state != M0_BOS_DONE);
-	M0_ASSERT( child->bo_sm.sm_state == M0_BOS_INIT);
+	M0_ASSERT(!parent->bo_set_addition_finished);
 
 	be_op_set_add(parent, child);
-	parent->bo_is_op_set = true;
 
 	m0_be_op_unlock(parent);
 	m0_be_op_unlock(child);
+	M0_LEAVE("parent=%p child=%p", parent, child);
+}
+
+M0_INTERNAL void m0_be_op_set_add_finish(struct m0_be_op *op)
+{
+	struct m0_be_op *child;
+
+	M0_ENTRY("op=%p", op);
+	M0_ASSERT(M0_IN(op->bo_kind, (M0_BE_OP_KIND_SET_AND,
+	                              M0_BE_OP_KIND_SET_OR)));
+	m0_be_op_lock(op);
+	M0_ASSERT(op->bo_set_triggered_by == NULL);
+	op->bo_set_addition_finished = true;
+	M0_ASSERT(!bos_tlist_is_empty(&op->bo_children));
+	/*
+	 * Concurrent additions to the list couldn't happen because of
+	 * op->bo_set_addition_finished.
+	 * Concurrent removals may happen.
+	 */
+	m0_tl_for(bos, &op->bo_children, child) {
+		m0_be_op_unlock(op);
+		/**
+		 * Lock order here should be the same as in m0_be_op_set_add()
+		 * and in be_op_done().
+		 */
+		m0_be_op_lock(child);
+		m0_be_op_lock(op);
+		/*
+		 * Concurrent removal could happen in be_op_done():
+		 * - for M0_BE_OP_KIND_SET_OR op called by a child of the op in
+		 *   the removal loop for op;
+		 * - for M0_BE_OP_KIND_SET_AND op when the last child of the op
+		 *   transitions to M0_BTS_DONE state.
+		 */
+		if (child->bo_parent != op) {
+			M0_LOG(M0_DEBUG, "concurrent removal of "
+			       "child=%p from op=%p happened", child, op);
+			m0_be_op_unlock(child);
+			m0_be_op_unlock(op);
+			return;
+		}
+		if (op->bo_set_triggered_by != NULL) {
+			M0_LOG(M0_DEBUG, "op=%p: had been triggered "
+			       "somewhere else", op);
+			m0_be_op_unlock(child);
+			m0_be_op_unlock(op);
+			return;
+		}
+		if (m0_be_op_is_done(child)) {
+			/*
+			 * The rest of the children for M0_BE_OP_KIND_SET_OR
+			 * op are removed in be_op_done().
+			 */
+			be_op_set_del(op, child);
+			/* extra parentheses to satisfy -Werror=parentheses */
```

Comment:
\*question\* We have a similar expression in `be_op_done` but it has no comments about `-Werror=parentheses`. Could we drop this comment? I guess we can freely assume that the reader knows that compilers would complain about mixing of different kinds of logical operators within a single expression without parentheses.